### PR TITLE
[Ready for review]Experimentor Rework

### PR DESCRIPTION
**Pull request by PidgeyThePirate**
Walked back the PR a bit to the state before I started adding stuff.

**What does this PR do:**
tl;dr: Reworks the Experimentor and the core gameplay loop around it to be slightly more engaging by adding more possible outcomes and an amount of control over those outcomes.

Summary:
- The Experimentor no longer takes any item, only the reworked strange objects (Unknown Tech) and their products.
- The most common new object (Prototype Tech) spawns in place of strange objects.
- Tech can be "unboxed" in the experimentor, revealing an item from a loot list with four rarity tiers (common, uncommon, etc.). The odds of gaining a very rare item are adjusted to occur roughly 1 in every 10-12 rounds.
- Other types of tech (Mysterious Tech and Alien Tech) are present: Mysterious tech can be found rarely in the maintenance tunnels. These types have higher base stats and a higher rarity multiplier, making better outcomes more common.
- Each Tech item can be tweaked in the Experimentor before unboxing to improve the stats of the resulting item. An undesirable item can be "recycled" in the experimentor for new tech at a rate of 2 items per one tech.

[Design Notes](https://docs.google.com/document/d/18wMcrupukvFLSRLp4ukFo_Y8M2TcayftWlKYDCa_hNU/edit?usp=sharing)

[New Loot Types with sprites by Drakeven (For after merge)](https://github.com/PidgeyThePirate/Paradise/tree/experimentor-rework-new-items)

[New Sprite samples (Not included in this PR; the reimplemented strange objects still use assembly icons until the next PR goes through)](https://imgur.com/a/y0CFfr5)

**Changelog:**
:cl:
add: New functions and potential loot for the experimentor!
del: Most of the old experimentor code.
wip: Loot lists are still extremely bare, as are the item icons.
experiment: basically this whole thing.
/:cl:

### CODE REVIEW

**Comment by TDSSS** [comment]
very interested in seeing the experimentor be more useful

**Comment by lolzyking** [comment]
Can we Add more ways to find Unknown tech than Just maint loot.

Perhaps a chance of randomly spawning in a cargo shipment, or found in mining, or being in an unopened locker in a department?

**Comment by PidgeyThePirate** [comment]
I was actually planning to have mysterious and alien tech show up rarely in lavaland and allowing prototype machines from cargo to be used as prototype tech. So I have those sorts of things in mind, though I'm trying to keep it all to a small set of files for the moment.

File: code/modules/research/experimentor/experimentor.dm [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+		return
+
+	if(disabled)
+		return
+
+	if(loaded_item)
+		to_chat(user, "<span class='warning'>The [src] is already loaded.</span>")
+		return
+
+	if(istype(O, /obj/item/unknown_tech) || (O != null && O.type in allowedobjects))
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: Wouldn't it be easier here to add the unknown_tech type to the allowedobjects list and do it all with one check?

Comment (PidgeyThePirate) [reply code/modules/research/experimentor/experimentor.dm]: Ah, good catch.

File: code/modules/research/experimentor/experimentor.dm [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+		if(!dropturf) //Failsafe to prevent the object being lost in the void forever.
+			dropturf = get_turf(src)
+		loaded_item.loc = dropturf
+		if(delete)
+			qdel(loaded_item)
+		loaded_item = null
+
+//Returns a string based on the level of the checked stat.
+/obj/machinery/r_n_d/experimentor/proc/getvaguestat(var/stat)
+	if (stat<21)
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: spacing (brackets should touch the if, applies to everything below too.

**Comment by Couls** [comment]
No sprites for the unknown techs yet?

File: code/modules/research/experimentor/experimentor.dm [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+	else
+		if(!loaded_item)
+			updateUsrDialog() //Set the interface to unloaded mode
+			to_chat(usr, "<span class='warning'>[src] is not currently loaded!</span>")
+			return
+		//It's harder to raise innovation.
+		var/adjusted_flex_cost = base_flex_cost - (round(T.stability/20,1)-2)
+		if(text2num(scantype) == SCANTYPE_POKE)
+			T.adjuststability(rand(2,10))
+			T.adjustinnovation(0-rand(0,6))
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: Why do 0 - something, instead of just putting -something there by itself?

File: code/modules/research/experimentor/experimentor.dm [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+			to_chat(usr, "<span class='notice'>The E.X.P.E.R.I-MENTOR adds a new coolant cell and some tubing to the [T.name].</span>")
+			playsound(loc, 'sound/items/welder.ogg', 50, 1)
+		if(text2num(scantype) == SCANTYPE_DISCOVER)
+			loaded_item = discover(T)
+			ejectItem()
+			to_chat(usr, "<span class='notice'>The E.X.P.E.R.I-MENTOR assembles the [T.name] into its new, final form!</span>")
+			playsound(src.loc, 'sound/items/rped.ogg', 50, 1)
+		if(text2num(scantype) == SCANTYPE_OBLITERATE)
+			T.reroll()
+			T.adjustflexibility(0-rand(adjusted_flex_cost*2+5, adjusted_flex_cost*3+5))
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: Seems to be the same as just going 5-rand(adjusted_flex_cost*2, adjusted_flex_coast*3), which reads easier, imo.

File: code/modules/research/experimentor/experimentor.dm [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+			playsound(src.loc, 'sound/machines/ping.ogg', 50, 0)
+			loaded_item = new/obj/item/unknown_tech/proto_tech()
+			spareparts -= 100
+			ejectItem()
+	src.updateUsrDialog()
+	return
+
+/obj/machinery/r_n_d/experimentor/proc/criticalfailure()
+	var/failuretype = rand(1,8)
+	switch(failuretype)
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: can't you just do
switch(rand(1,8)) here?

File: code/modules/research/experimentor/experimentor.dm [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+			var/turf/MT = get_turf(M)
+			if(MT)
+				visible_message("<span class='danger'>The coolant cells burst violently!</span>")
+				var/obj/item/projectile/magic/fireball/FB = new /obj/item/projectile/magic/fireball(start)
+				FB.original = MT
+				FB.current = start
+				FB.yo = MT.y - start.y
+				FB.xo = MT.x - start.x
+				FB.fire()
+		if(6)
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: Not a code comment, but I'm kind of disappointed the experimentor can no longer spawn angry space bears or other mobs like that.

Comment (PidgeyThePirate) [reply code/modules/research/experimentor/experimentor.dm]: You're not wrong, this is a serious issue that requires careful handling.

File: code/modules/research/experimentor/experimentor.dm [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+					spawn(0)
+						AM.throw_at(src,10,1)
+
+
+
+// Responsible for generating and returning a new functional object to replace the old box.
+/obj/machinery/r_n_d/experimentor/proc/discover(var/obj/item/unknown_tech/T)
+	//Rarity Roll - Starts common. Better tiers overwrite if they crop up.
+	var/rarity = RARITY_COMMON
+	if (prob(min(((T.uncommon_weighting*T.innovation/100+1)*T.uncommon_base), 100)))
```
Comment (TDSSS) [inline code/modules/research/experimentor/experimentor.dm]: after some quick testing, calling prob() with a number greater than 100 doesn't change anything, it just treats it as a 100% chance, so you can remove all the mins.

File: code/modules/research/experimentor/loot/_loot_definer.dm [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+		loot_item.desc = generateDescription()
+	return loot_item
+
+
+///////////////////I. LOOT LIST////////////////////
+// Defines the loot lists for the experimentor
+// by category then by rarity.
+
+/obj/experimentor/loot_definer/proc/findItemType()
+    switch (item_category)
```
Comment (TDSSS) [inline code/modules/research/experimentor/loot/_loot_definer.dm]: What's the purpose of this switch statement?

File: code/modules/research/experimentor/loot/_loot_definer.dm [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+/obj/experimentor/loot_definer/proc/definitionByType()
+    // Switch here mostly for our benefit to break up categories.
+    // Try to keep your loot and device types in alphabetical order.
+    switch(item_category)
+        if("Device")
+            switch(item_type)
+                if("clean")
+                    loot_keywords.Add("destruction")
+                    return new/obj/item/discovered_tech(stability, potency, raritylevel, item_type)
```
Comment (TDSSS) [inline code/modules/research/experimentor/loot/_loot_definer.dm]: those all seem to return the same thing, so you have a lot of duplicate reutrns here. Why not simply have a single
return new/obj/item... at the end, after the ifs?

File: code/modules/research/experimentor/loot/_loot_definer.dm (hunk 8 above)
Comment (PidgeyThePirate) [reply code/modules/research/experimentor/loot/_loot_definer.dm]: Expandability. The system just has results of type "Device" for now, but eventually there will be more than one Type (loot list) to draw from.

File: code/modules/research/experimentor/loot/_loot_definer.dm (hunk 9 above)
Comment (PidgeyThePirate) [reply code/modules/research/experimentor/loot/_loot_definer.dm]: I probably should do that, but this set up I did as an example; just because they all draw from the same object now doesn't mean they will always do so in that loot list.

File: code/modules/research/experimentor/loot/_loot_definer.dm [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    var/loot_named
+    var/loot_described
+    // Keywords used to choose names for the second and third parts of the generated name.
+    // When defining new keywords, try to use ones that fit in the auto-generated descriptions.
+    var/list/loot_keywords
+    // Unique Names - If instantiated in the definition, will use one of the contained names instead of generating the third name.
+    var/list/unique_names
+
+// Returns a finished item from the stats supplied to the experimentor.
+/obj/experimentor/loot_definer/proc/define(var/stability_in, var/potency_in, var/base_name, var/rare_level, var/itemcategory)
```
Comment (TDSSS) [inline code/modules/research/experimentor/loot/_loot_definer.dm]: should this really be an obj instead of a datum?

Comment (PidgeyThePirate) [reply code/modules/research/experimentor/loot/_loot_definer.dm]: probably not.

Comment (Couls) [reply code/modules/research/experimentor/loot/_loot_definer.dm]: Might be better as a datum yeah,

**Comment by TDSSS** [comment]
Okay, having reviewed the entire code, it seems very reliant on ifs with strings. I am not sure thats the best/cleanest way to do what you want, tbh. It looks pretty messy and a pain to expand to me.

**Comment by PidgeyThePirate** [comment]
What would you recommend instead?

File: code/modules/research/experimentor/experimentor.dm (hunk 1 above)
Comment (PidgeyThePirate) [reply code/modules/research/experimentor/experimentor.dm]: I figured out why. Adding unknown_tech to the allowed objects list allows unknown_tech and ONLY unknown_tech, not its descendants. Can you do istype efficiently with a list?

**Comment by TDSSS** [comment]
I'll give you an example of how your code currently functions, and then how I'd do it.

Let's say we want to add a new kind of device, for the purpose of this example we'll call it the "coffeemaker". With the code as you have written it, we need to add a new keyword in loot_definer.dm, adding "coffeemaker" to the pick list. Then later we have to go to definitionByType(), add a new line there as well to add some keywords, then you need to define a new proc named after it, etc.

In other words, you have to modify several different files in different places to add something new, and you rely on strings as keywords. This isn't very object oriented at all.

Instead, I'd do something like put all the information in one object. Let's say we make a datum somewhat like this, using the teleport object as an example:
```
datum/tech_info/teleport
//we give the tech_info class a bunch of stuff we need that we can override to use later. for example:
datum/tech_info/teleport/choosecooldown()
    return round(rand(150,300)*(1.4-0.2*raritylevel)*(1.4-0.2*(stability/20)),1)

//we also give it a list of keywords, then when generating the object, we simply add all those to our end result:
loot_keywords = list("teleportation", "bluespace")
name_keywords1 = list("distorting", "gravitic")
//...

//and rather than the  proc/teleport we simply override some basic use proc:
datum/tech_info/teleport/use_tech(...
```

I hope that makes sense. All the actual code's written off the cuff, might not work exactly like that. But am I getting across the general idea?

**Comment by PidgeyThePirate** [comment]
I see, so you can describe a type of tech in one place instead of like... 3? I can work with that.

**Comment by TDSSS** [comment]
Honestly, you can probably skip the datum step and just chose the 'end result' object directly.

**Comment by PidgeyThePirate** [comment]
Right, and then give the keywords to the object directly. Though since they're all the same object I'd probably have to populate the keywords in the object itself, which is only a small improvement.

**Comment by TDSSS** [comment]
Why _are_ they all the same object in the first place?

**Comment by PidgeyThePirate** [comment]
Because that's how the original strange objects did it. That's all, really.

**Comment by PidgeyThePirate** [comment]
This should look better. What do you think?

**Comment by TDSSS** [comment]
Looks better to me, though a maint can probably give more authorative statements on that.

As for the tech levels, the way I see it, you can in theory get a origin_tech of 8, right? That will have some gameplay implications, though I assume very rare level items are going to be vanishingly rare and usually not worth deconstructing, so it balances out.

**Comment by PidgeyThePirate** [comment]
Yeah, you'll most often get rare at most and those will max out at 6, which is more reasonable. Very Rare I've tried to balance so it crops up 1 out of 10 rounds if you open up 10 boxes per round including  one mysterious tech, and as far as tech levels go you'd have to want to burn your very rare in the analyser to get the tech (which isn't a problem now, but once the loot starts getting actually desirable...)

**Comment by Fox-McCloud** [comment]
Question.

How much of this will be relevant or useful if we implement techwebs? 

I'd hate for you to refactor all of this for it to be useless a month or so down teh road---or for it to be the "no, we can't implement techwebs because of the ExperiMENTOR refactor!"

**Comment by PidgeyThePirate** [comment]
I don't know much about techwebs, but I highly doubt that it will be relevant since it has little to do with  tech level production.

**Comment by PidgeyThePirate** [comment]
I got around to having a look at TG's experimentor, and I don't foresee any issues combining the two functions. My rework essentially replaces the relic functions of the experimentor and scraps the rest, so you should just be able to plug in what tg has with a little UI tinkering. Maybe have it switch to Discover Tech mode only when it's loaded with a mysterious tech item.

File: code/modules/research/experimentor/experimentor.dm [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+//
+//
+
+/obj/machinery/r_n_d/experimentor
+	name = "E.X.P.E.R.I-MENTOR"
+	icon = 'icons/obj/machines/heavy_lathe.dmi'
+	desc = "A marvel of technology, the onboard AI is capable of turning compatible material into brand-new contraptions... sometimes."
+	icon_state = "h_lathe"
+	var/list/allowedobjects = list(/obj/item/discovered_tech, /obj/item/unknown_tech)
+	density = 1
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: Best practice is to use TRUE and FALSE for boolean values

File: code/modules/research/experimentor/experimentor.dm [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+	anchored = 1
+	use_power = IDLE_POWER_USE
+	var/base_flex_cost = 10
+	var/precise_scanner = 0
+	var/spareparts = 0
+	var/datum/experimentor/loot_definer/LootDefiner = new/datum/experimentor/loot_definer()
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: can just use new() here. It will default to the specified type on the left hand side.

File: code/modules/research/experimentor/experimentor.dm [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+/obj/machinery/r_n_d/experimentor/RefreshParts()
+	var/part_rating_count = -4
+	base_flex_cost = 10
+	for(var/obj/item/stock_parts/manipulator/M in component_parts)
+		part_rating_count += M.rating
+	for(var/obj/item/stock_parts/micro_laser/M in component_parts)
+		part_rating_count += M.rating
+	for(var/obj/item/stock_parts/scanning_module/M in component_parts)
+		precise_scanner = M.rating
+	base_flex_cost -= round(part_rating_count/3,1)
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: spacing between the operators and the ,

File: code/modules/research/experimentor/experimentor.dm [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+	if(shocked)
+		shock(user,50)
+
+	if(default_deconstruction_screwdriver(user, "h_lathe_maint", "h_lathe", O))
+		return
+
+	if(exchange_parts(user, O))
+		return
+
+	if(panel_open && istype(O, /obj/item/crowbar))
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: Should be moved to /obj/machinery/r_n_d/experimentor/crowbar_act(...)

File: code/modules/research/experimentor/experimentor.dm [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+		part_rating_count += M.rating
+	for(var/obj/item/stock_parts/scanning_module/M in component_parts)
+		precise_scanner = M.rating
+	base_flex_cost -= round(part_rating_count/3,1)
+
+/obj/machinery/r_n_d/experimentor/attackby(obj/item/O, mob/user, params)
+	if(shocked)
+		shock(user,50)
+
+	if(default_deconstruction_screwdriver(user, "h_lathe_maint", "h_lathe", O))
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: Should be moved to /obj/machinery/r_n_d/experimentor/screwdriver_act(...)

File: code/modules/research/experimentor/experimentor.dm [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+		return
+
+	if(disabled)
+		return
+
+	if(loaded_item)
+		to_chat(user, "<span class='warning'>The [src] is already loaded.</span>")
+		return
+
+	if(O != null && (istype(O, /obj/item/unknown_tech) || istype(O, /obj/item/discovered_tech)))
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: `O != null` can be written as `!O` here

Comment (farie82) [reply code/modules/research/experimentor/experimentor.dm]: istype will also check if it's null by the way. `!O && istype(O, path)` is the same as `istype(O,path)`

File: code/modules/research/experimentor/experimentor.dm [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+		if(!user.drop_item())
+			return
+		loaded_item = O
+		O.loc = src
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: use O.forceMove(src)

File: code/modules/research/experimentor/experimentor.dm [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+		return
+	return
+
+/obj/machinery/r_n_d/experimentor/default_deconstruction_crowbar(obj/item/O)
+	ejectItem()
+	..(O)
+
+/obj/machinery/r_n_d/experimentor/proc/ejectItem(delete=FALSE)
+	if(loaded_item)
+		var/turf/dropturf = get_turf(pick(view(1,src)))
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: spacing between the '1,' and 'src'

File: code/modules/research/experimentor/experimentor.dm [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+		O.loc = src
+		to_chat(user, "<span class='notice'>You add the [O.name] to the E.X.P.E.R.I-MENTOR.</span>")
+		flick("h_lathe_load", src)
+	// Checks the allowedobjects list to determine whether the object is an unboxed one.
+	else
+		to_chat(user, "<span class='warning'>The E.X.P.E.R.I-MENTOR refuses the [O.name].</span>")
+		return
+	return
+
+/obj/machinery/r_n_d/experimentor/default_deconstruction_crowbar(obj/item/O)
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: this can just become crowbar_act. Just add the user to the params

File: code/modules/research/experimentor/experimentor.dm [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+		if(!dropturf) //Failsafe to prevent the object being lost in the void forever.
+			dropturf = get_turf(src)
+		loaded_item.loc = dropturf
+		if(delete)
+			qdel(loaded_item)
+		loaded_item = null
+
+//Returns a string based on the level of the checked stat.
+/obj/machinery/r_n_d/experimentor/proc/getvaguestat(var/stat)
+	if(stat<21)
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: spacing in here as well

File: code/modules/research/experimentor/experimentor.dm [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+		var/turf/dropturf = get_turf(pick(view(1,src)))
+		if(!dropturf) //Failsafe to prevent the object being lost in the void forever.
+			dropturf = get_turf(src)
+		loaded_item.loc = dropturf
+		if(delete)
+			qdel(loaded_item)
+		loaded_item = null
+
+//Returns a string based on the level of the checked stat.
+/obj/machinery/r_n_d/experimentor/proc/getvaguestat(var/stat)
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: `var/stat` can just be `stat` in params. var is not required

File: code/modules/research/experimentor/experimentor.dm [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+		dat += "<b>Nothing loaded.</b>"
+	dat += "<br><a href='byond://?src=[UID()];function=refresh'>Refresh</A><br>"
+	dat += "<br><a href='byond://?src=[UID()];close=1'>Close</A><br></center>"
+	var/datum/browser/popup = new(user, "experimentor","Experimentor", 500, 700, src)
+	popup.set_content(dat)
+	popup.open()
+	onclose(user, "experimentor")
+
+/obj/machinery/r_n_d/experimentor/Topic(href, href_list)
+	var/obj/item/unknown_tech/T = locate(href_list["item"]) in src
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: this var declaration can be lowered till you are sure it is required (small optimization)

File: code/modules/research/experimentor/experimentor.dm [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+			to_chat(usr, "<span class='warning'>[src] is not currently loaded!</span>")
+			return
+		//It's harder to raise innovation.
+		var/adjusted_flex_cost = base_flex_cost - (round(T.stability/20,1)-2)
+		if(text2num(scantype) == SCANTYPE_POKE)
+			T.adjuststability(rand(2,10))
+			T.adjustinnovation(-rand(0,6))
+			T.adjustflexibility(-rand(adjusted_flex_cost, adjusted_flex_cost + 5))
+			to_chat(usr, "<span class='notice'>Mechanical arms carefully grease and work the [T.name] as lasers prune away excess weight.</span>")
+			playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: src is not required here. Same for the ones below

File: code/modules/research/experimentor/experimentor.dm [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+#undef SCANTYPE_POKE
+#undef SCANTYPE_IRRADIATE
+#undef SCANTYPE_GAS
+#undef SCANTYPE_HEAT
+#undef SCANTYPE_COLD
+#undef SCANTYPE_OBLITERATE
+#undef SCANTYPE_DISCOVER
+#undef SCANTYPE_RECYCLE
+
+#undef RARITY_COMMON
```
Comment (farie82) [inline code/modules/research/experimentor/experimentor.dm]: Instead of undefing them here I'd suggest making them start with say EXPERIMENTOR_....
This way you don't have to redef them in other files. Might also want to move the defines to the _DEFINE folder since they are used in multiple files

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+	icon_state = pick(iconlist)
+	techLevels = raritylevel*2
+	box_type = boxtype
+	initialize()
+
+/obj/item/discovered_tech/attack_self(mob/user)
+	if(cooldown)
+		to_chat(user, "<span class='warning'>The [src] does not react!</span>")
+		return
+	else if(src.loc == user)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: src implied

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+	var/extra_data
+	var/box_type
+
+	var/list/keywords = list("clowns")
+	var/use_generated_names = TRUE
+	var/use_generated_descriptions = TRUE
+
+	var/techLevels
+
+/obj/item/discovered_tech/New(var/stability_in, var/potency_in, var/rare_level, var/boxtype)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: var implied

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 27]
```diff
@@ -0,0 +1,236 @@
...
+		origin_tech += "abductor=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+
+/obj/item/discovered_tech/proc/itemproc(mob/user)
+	return
+
+/obj/item/discovered_tech/proc/warn_admins(mob/user, ItemType, priority = 1)
+	var/turf/T = get_turf(src)
+	var/log_msg = "[ItemType] experimentor tech used by [key_name(user)] in ([T.x],[T.y],[T.z])"
+	if(priority) //For truly dangerous relics that may need an admin's attention. BWOINK!
+		message_admins("[ItemType] experimentor tech activated by [key_name_admin(user)] in ([T.x], [T.y], [T.z] - <A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[T.x];Y=[T.y];Z=[T.z]'>JMP</a>)",0,1)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: use `ADMIN_COORDJMP` instead of the manual typing

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 28]
```diff
@@ -0,0 +1,236 @@
...
+	..()
+	origin_tech += "materials=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
+	keywords = list("destruction", "light")
+
+/obj/item/discovered_tech/smokebomb/itemproc(turf/where)
+	visible_message("<span class='notice'>The [src] belches forth a cloud of smoke!</span>")
+	var/datum/effect_system/smoke_spread/smoke = new
+	smoke.set_up(potency/25,0, where, 0)
+	smoke.start()
+	return
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: return not required

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 29]
```diff
@@ -0,0 +1,236 @@
...
+// Floof Cannon
+/obj/item/discovered_tech/floofcannon/initialize()
+	..()
+	origin_tech += "biotech=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	origin_tech += "bluespace=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
+	keywords = list("animals", "replication")
+	extra_data = pick(/mob/living/simple_animal/pet/corgi, /mob/living/simple_animal/pet/cat, /mob/living/simple_animal/pet/fox, /mob/living/simple_animal/mouse, /mob/living/simple_animal/pet/pug, /mob/living/simple_animal/lizard, /mob/living/simple_animal/diona, /mob/living/simple_animal/butterfly, /mob/living/carbon/human/monkey)
+
+/obj/item/discovered_tech/floofcannon/itemproc(mob/user)
+	playsound(src.loc, "sparks", rand(25,50), 1)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: src implied

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 30]
```diff
@@ -0,0 +1,236 @@
...
+	origin_tech += "combat=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	origin_tech += "plasmatech=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	if(raritylevel>=RARITY_RARE)
+		origin_tech += "toxins=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
+	keywords = list("destruction", "sound")
+
+/obj/item/discovered_tech/explosion/itemproc(mob/user)
+	to_chat(user, "<span class='danger'>[src] begins to heat up!</span>")
+	spawn(rand(max(stability,35),100))
+		if(src.loc == user)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: src implied. Last one I'll comment on

File: code/modules/research/experimentor/loot/device_exploot.dm (hunk 27 above)
Comment (farie82) [reply code/modules/research/experimentor/loot/device_exploot.dm]: message_admins will probably also be to much for logging. Admins will have to give you pointers which logging is better to use though

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 31]
```diff
@@ -0,0 +1,236 @@
...
+	var/obj/item/grenade/flashbang/CB = new/obj/item/grenade/flashbang(get_turf(user))
+	CB.range = rand(stability/25, max(potency/10, 4))
+	CB.prime()
+	warn_admins(user, "Flashbang")
+
+// Rapid Duplicator
+/obj/item/discovered_tech/rapidDuplicator/initialize()
+	..()
+	origin_tech += "magnets=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	origin_tech += "engineering=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: missing ; at the end of the string. Best to make this into a proc and use it for all techs since you use it so often

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 32]
```diff
@@ -0,0 +1,236 @@
...
+	var/list/dupes = list()
+	var/counter
+	var/max = rand(stability/20, max(potency/5, 5))
+	for(counter = 1; counter < max; counter++)
+		var/obj/item/discovered_tech/R = new src.type(get_turf(src))
+		R.icon_state = icon_state
+		R.name = name
+		R.desc = desc
+		R.throwforce = ((potency/10)*raritylevel)+1
+		dupes |= R
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: |= is not needed. |= means it will uniquely add. So if it's already in the list it won't add it. But this also means it's more expensive than += which just adds it. Or you can do dupes.Add(R)

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 33]
```diff
@@ -0,0 +1,236 @@
...
+		spawn()
+			R.throw_at(pick(oview(4+potency/20,get_turf(src))),10,1)
+	counter = 0
+	spawn(rand(stability/10,stability))
+		for(counter = 1; counter <= dupes.len; counter++)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: can be replaced by a foreach:
```
for(var/thingy in dupes)
    qdel(thingy)
```

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 34]
```diff
@@ -0,0 +1,236 @@
...
+			var/obj/item/discovered_tech/R = dupes[counter]
+			qdel(R)
+
+// Teleporter
+/obj/item/discovered_tech/teleport/initialize()
+	..()
+	origin_tech += "bluespace=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	origin_tech += "powerstorage=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
+	keywords = list("teleportation", "bluespace")
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: best to put this in the item definition itself. Not in the initialize() or New().
If it's placed there then it will use less memory and performance etc. Same for the other items above and below

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 35]
```diff
@@ -0,0 +1,236 @@
...
+
+/obj/item/discovered_tech/teleport/itemproc(mob/user)
+	to_chat(user, "<span class='notice'>The [src] begins to vibrate!</span>")
+	spawn(rand((15-stability/10),30-stability/5))
+		var/turf/userturf = get_turf(user)
+		if(src.loc != user || is_teleport_allowed(userturf.z) == FALSE)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: == FALSE can be placed with `!statement` so in this case `!is_teleport_allowed(userturf.z)`

File: code/modules/research/experimentor/loot/device_exploot.dm [hunk 36]
```diff
@@ -0,0 +1,236 @@
...
+/obj/item/discovered_tech/massSpawner/initialize()
+	..()
+	origin_tech += "biotech=[1+raritylevel+rand(raritylevel,raritylevel+1)];"
+	origin_tech += "bluespace=[1+raritylevel+rand(raritylevel,raritylevel+1)]"
+	keywords = list("animals", "replication")
+
+/obj/item/discovered_tech/massSpawner/itemproc(mob/user)
+	var/message = "<span class='danger'>[src] begins to shake, and in the distance the sound of rampaging animals arises!</span>"
+	visible_message(message)
+	to_chat(user, message)
```
Comment (farie82) [inline code/modules/research/experimentor/loot/device_exploot.dm]: Will duplicate the message since visible_message should already message the user

File: code/modules/research/experimentor/unknown_techs.dm [hunk 37]
```diff
@@ -0,0 +1,162 @@
...
+	icon_state = pick(iconlist)
+	containedtype = pick(typelist)
+	stability = NormDistRand(stability_standardeviation, stability_mean)
+	potency = NormDistRand(potency_standardeviation, potency_mean)
+	innovation = NormDistRand(innovation_standardeviation, innovation_mean)
+	flexibility = rand(flexibility_min, flexibility_max)
+
+/obj/item/unknown_tech/proc/adjuststability(var/statchange)
+	stability += statchange
+	if(stability > 100)
```
Comment (farie82) [inline code/modules/research/experimentor/unknown_techs.dm]: can use clamp for these checks `clamp(value, min, max)`

**Comment by PidgeyThePirate** [comment]
Oh wow, haven't touched this one in ages. All good points (and largely ones that I have learned in projects that came after this, I think).

When I get my computer back I'll have to sit down for a revision here.
